### PR TITLE
Add font and image types

### PR DESCRIPTION
**Pull request by dfrg**
This adds a new `Blob` type that is used as a general shared buffer with a generated unique identifier and then builds types for font and image resources on top of that.

This should enable initial support for a proper glyph run API in vello.

It also adds an `Image` variant to the `Brush` type and reworks gradients for better code reuse, cleaning up and simplifying the API a bit.

### CODE REVIEW

File: src/gradient.rs [hunk 1]
```diff
@@ -56,131 +72,132 @@ impl From<(f32, Color)> for ColorStop {
...
 
-/// Definition of a gradient that transitions between two or more colors along
-/// a line.
-#[derive(Clone, PartialEq, Default, Debug)]
-pub struct LinearGradient {
-    pub start: Point,
-    pub end: Point,
+/// Properties for the supported gradient types.
+#[derive(Clone, PartialEq, Debug)]
+pub enum GradientKind {
```
Comment (raphlinus) [inline src/gradient.rs:77]: The word "kind" seems slightly wrong, as I'd expect it to be a simple enum. But I can't immediately think of a better word, and that will be clear from usage anyway.

File: src/gradient.rs [hunk 2]
```diff
@@ -56,131 +72,132 @@ impl From<(f32, Color)> for ColorStop {
...
+    /// Gradient that transitions between two or more colors along a line.
+    Linear {
+        /// Starting point.
+        start: Point,
+        /// Ending point.
+        end: Point,
+    },
+    /// gradient that transitions between two or more colors that radiate from an origin.
```
Comment (raphlinus) [inline src/gradient.rs:85]: Capitalization not consistent

File: src/gradient.rs [hunk 3]
```diff
@@ -56,131 +72,132 @@ impl From<(f32, Color)> for ColorStop {
...
+        start_angle: f32,
+        /// End angle of the sweep, counter-clockwise of the x-axis.
+        end_angle: f32,
+    },
+}
+
+/// Definition of a gradient that transitions between two or more colors.
+#[derive(Clone, PartialEq, Debug)]
+pub struct Gradient {
+    /// Kind of the gradient.
```
Comment (raphlinus) [inline src/gradient.rs:111]: Maybe say "Kind and details" here?

File: src/image.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use super::{Blob, Extend};
+
+/// Defines the pixel format of an image.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum Format {
+    /// 32-bit RGBA with 8-bit channels.
```
Comment (raphlinus) [inline src/image.rs:22]: Really going all-out here supporting all the different formats, are we?

**Review by raphlinus — APPROVED** [review]
Overall looks good.

There's such a huge potential rathole to get color fully correct, but I think erring on the side of simplicity is the right thing for now.

File: src/blob.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
+// Copyright 2022 The peniko authors.
```
[the remark this answers is not in the document]
Comment (dfrg) [reply src/blob.rs:1]: Very happy to reduce the license noise. I was trying to keep it consistent with vello which still uses the full license text outside of the shaders. I'll update the files.

File: src/blob.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Also licensed under MIT license, at your choice.
+
+use std::fmt;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Weak};
+
+/// Shared data with an associated unique identifier.
```
[the remark this answers is not in the document]
Comment (dfrg) [reply src/blob.rs:21]: Sure, I can definitely see that being useful. Will add it.

File: src/color.rs [hunk 7]
```diff
@@ -140,10 +138,18 @@ impl Color {
...
 
+    /// Returns the color with the alpha component multiplied by the specified
+    /// factor.
+    pub fn with_alpha_factor(self, alpha: f32) -> Self {
+        let mut result = self;
+        result.a = ((result.a as f32) * alpha) as u8;
+        result
+    }
+
     /// Premultiplies the color by the alpha component.
```
[the remark this answers is not in the document]
Comment (dfrg) [reply src/color.rs:153]: Yes, I don't think anything actually depends on this, so I'm happy to just remove these functions. I don't know why I originally added them.

File: src/gradient.rs (hunk 2 above)
Comment (dfrg) [reply src/gradient.rs:85]: Doh

File: src/gradient.rs (hunk 3 above)
Comment (dfrg) [reply src/gradient.rs:111]: Will do.

File: src/image.rs (hunk 4 above)
Comment (dfrg) [reply src/image.rs:22]: I think this still fits nicely within the zero, one, many rule? :)

File: src/gradient.rs (hunk 1 above)
Comment (dfrg) [reply src/gradient.rs:77]: I don't disagree, but the pattern of having common fields in `S` and variant dependent fields `SKind` pops up all over the rustc codebase so I think I internalized it?

**Pull request merged**